test/integration/mqtt_paho_binding/mqtt_test.go

```diff
@@ -0,0 +1,99 @@
+package mqtt_paho_binding
+
+import (
+	"context"
+	"net"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/eclipse/paho.golang/paho"
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/cloudevents/sdk-go/protocol/mqtt_paho/v2"
+	"github.com/cloudevents/sdk-go/v2/binding"
+	"github.com/cloudevents/sdk-go/v2/event"
+	bindings "github.com/cloudevents/sdk-go/v2/protocol"
+	"github.com/cloudevents/sdk-go/v2/protocol/test"
+
+	. "github.com/cloudevents/sdk-go/v2/binding/test"
+	. "github.com/cloudevents/sdk-go/v2/test"
+)
+
+func TestSendStructuredMessageToStructured(t *testing.T) {
+	ctx := context.Background()
+	s, r, close := testSenderReceiver(t, ctx)
+	defer close()
+
+	EachEvent(t, Events(), func(t *testing.T, eventIn event.Event) {
+		eventIn = ConvertEventExtensionsToString(t, eventIn)
+		in := MustCreateMockStructuredMessage(t, eventIn)
+
+		test.SendReceive(t, binding.WithPreferredEventEncoding(ctx, binding.EncodingStructured), in, s, r,
+			func(out binding.Message) {
+				eventOut := MustToEvent(t, ctx, out)
+				assert.Equal(t, binding.EncodingStructured, out.ReadEncoding())
+				AssertEventEquals(t, eventIn, ConvertEventExtensionsToString(t, eventOut))
+			})
+	})
+}
+
+func TestSendBinaryMessageToBinary(t *testing.T) {
+	ctx := context.Background()
+	s, r, close := testSenderReceiver(t, ctx)
+	defer close()
+
+	EachEvent(t, Events(), func(t *testing.T, eventIn event.Event) {
+		eventIn = ConvertEventExtensionsToString(t, eventIn)
+		in := MustCreateMockBinaryMessage(eventIn)
+		test.SendReceive(t, binding.WithPreferredEventEncoding(ctx, binding.EncodingBinary), in, s, r,
+			func(out binding.Message) {
+				eventOut := MustToEvent(t, ctx, out)
+				assert.Equal(t, binding.EncodingBinary, out.ReadEncoding())
+				eventOut = ConvertEventExtensionsToString(t, eventOut)
+
+				AssertEventEquals(t, eventIn, ConvertEventExtensionsToString(t, eventOut))
+			})
+	})
+}
+
+func testSenderReceiver(t testing.TB, ctx context.Context) (bindings.Sender, bindings.Receiver, func()) {
+	topicName := "test-ce-client-" + uuid.New().String()
+
+	broker := os.Getenv("TEST_MQTT_BROKER")
+	if broker == "" {
+		broker = "127.0.0.1:1883"
+	}
+	senderConn, err := net.Dial("tcp", broker)
+	require.NoError(t, err)
+	cp := &paho.Connect{
+		KeepAlive:  30,
+		CleanStart: true,
+	}
+
+	s, err := mqtt_paho.New(ctx, &paho.ClientConfig{
+		Conn: senderConn,
+	}, cp, topicName, []string{topicName}, 0, false)
+	require.NoError(t, err)
+	require.NotNil(t, s)
+
+	receiverConn, err := net.Dial("tcp", broker)
+	require.NoError(t, err)
+	r, err := mqtt_paho.New(ctx, &paho.ClientConfig{
+		Conn: receiverConn,
+	}, cp, topicName, []string{topicName}, 0, false)
+	require.NoError(t, err)
+	require.NotNil(t, r)
+
+	go func() {
+		require.NoError(t, r.OpenInbound(context.TODO()))
+	}()
+
+	time.Sleep(5 * time.Millisecond) // let the receiver ready
```

Comment:
you can avoid this (anti-pattern) and the issue with failing in a goroutine above using channels and a ctx with timeout. please rewrite accordingly 🙏🏻